File: frontend/src/features/conversation/reportResolutions.ts
import { assessmentFinding } from "./assessment";

const statusLabels: Record<string, string> = { resolved: "已修正", unresolved: "仍存在", superseded: "要求已撤销或替换", unknown: "未知" };

/** 只读取普通对象，旧报告和无效字段不推断为已解决。 */
function record(value: unknown): Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : {};
}

/** 固定轮次和组号仅接受正整数。 */
function positive(value: unknown): number | null {
  return typeof value === "number" && Number.isSafeInteger(value) && value > 0 ? value : null;
}

/** 将独立的问题现状及逐组依据映射为视图，不覆盖历史失败评分。 */
export function reportResolutions(result: Record<string, unknown> | null) {
  const metadata = record(result?.report);
  const available = metadata.resolutionVersion === 1 && metadata.resolutionComplete === true && Array.isArray(metadata.issueResolutions);
  const items = (available && Array.isArray(metadata.issueResolutions) ? metadata.issueResolutions : []).map((raw: unknown) => {
    const item = record(raw);
    const reviews = (Array.isArray(item.reviews) ? item.reviews : []).map((rawReview: unknown) => {
      const review = record(rawReview);
      const valid = review.valid === true;
      return { index: positive(review.reviewIndex), valid,
        state: valid && typeof review.status === "string" ? statusLabels[review.status] ?? "未知" : "无有效结论",
        reason: typeof review.reason === "string" ? review.reason : "未记录理由",
        references: valid ? assessmentFinding({ evidence_refs: review.evidence }).references : [] };
    });
    return { id: typeof item.issueId === "string" ? item.issueId : "未知问题",
      description: typeof item.description === "string" ? item.description : "未记录问题说明",
      throughTurn: positive(item.throughTurn),
      state: typeof item.status === "string" ? statusLabels[item.status] ?? "未知" : "未知",
      reason: typeof item.reason === "string" ? item.reason : "未记录理由", reviews };
  });
  return { available, items };
}

File: backend/app/services/multiturn/report_resolutions.py
"""历史评分完成后独立判断问题现状，结果不参与原分数计算。"""
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker

from app.adapters.base import ModelClient
from app.models.conversation import ConversationJudgeRun
from app.services.multiturn.issue_resolution import IssueResolution, aggregate_resolution
from app.services.multiturn.assessment_batches import batch_count
from app.services.multiturn.resolution_store import ResolutionStore, execute_resolution
from app.services.multiturn.store import ConversationError
from app.services.token_quota_service import TokenQuotaExceededError


async def resolve_report_issues(sessions: async_sessionmaker[AsyncSession], job_id: int, owner: int,
                               client: ModelClient, *, output_tokens: int) -> None:
    """逐问题执行三组独立状态判断并一次保存汇总，重复作业不会再次发送。"""
    store = ResolutionStore()
    async with sessions() as db:
        job = await store._job(db, job_id, owner)
        if job.input_json["report"].get("resolutionComplete"):
            return
        rows = (await db.scalars(select(ConversationJudgeRun).where(ConversationJudgeRun.assessment_id == job_id)
            .order_by(ConversationJudgeRun.run_index))).all()
        expected = 3 * batch_count(job.input_json)
        if {row.run_index for row in rows} != set(range(1, expected + 1)) or any(row.status == "pending" for row in rows):
            raise ConversationError("resolution_state_conflict", "三组历史评审尚未完整保存，不能汇总问题状态", 409)
        failures = {item["id"] for row in rows if row.status == "provisional"
            for item in (row.result_json or {}).get("items", []) if item.get("applicability") == "applicable"
            and ((item.get("rating") is not None and item["rating"] < 3) or item.get("passed") is False)}
        checks = {check["id"]: check for check in job.input_json["packet"]["checks"]}
        through_turn = job.through_turn
    results: list[dict[str, object]] = []
    for issue_id in sorted(failures):
        entry: dict[str, object] = {"issueId": issue_id, "description": checks[issue_id]["description"],
            "throughTurn": through_turn, "status": "unknown", "reason": "尚未取得一致状态", "reviews": []}
        try:
            async with sessions() as db:
                packet = await store.packet(db, job_id, owner, issue_id)
        except ConversationError as error:
            if error.code not in ("resolution_failure_disputed", "resolution_anchor_missing"):
                raise
            entry.update(reason=str(error), errorCode=error.code)
            results.append(entry)
            continue
        except ValueError as error:
            entry.update(reason=str(error), errorCode="resolution_material_unavailable")
            results.append(entry)
            continue
        runs: list[IssueResolution | None] = []
        reviews: list[dict[str, object]] = []
        for index in range(1, 4):
            try:
                result = await execute_resolution(sessions, job_id, owner, issue_id, index, client, output_tokens=output_tokens)
            except TokenQuotaExceededError:
                entry.update(reason="状态评审额度不足，未执行的组保持未知", errorCode="resolution_quota_exceeded")
                reviews.extend({"reviewIndex": pending, "valid": False, "status": "unknown",
                    "reason": "额度不足，本组未执行", "errorCode": "resolution_not_run", "evidence": []}
                    for pending in range(index, 4))
                break
            runs.append(result.result)
            reviews.append({"reviewIndex": index, "valid": result.result is not None,
                "status": result.result.status if result.result else "unknown",
                "reason": result.result.reason if result.result else "本组未获得有效结论",
                "errorCode": result.error_code,
                "evidence": [ref.model_dump(mode="json") for ref in result.result.evidence] if result.result else []})
        completed = len(runs) == 3
        entry["status"] = aggregate_resolution(tuple(runs), packet) if completed else "unknown"
        entry["reviews"] = reviews
        if completed:
            entry["reason"] = "有效组结论一致" if entry["status"] != "unknown" else "状态未知、有效组不足或组间分歧"
        results.append(entry)
    async with sessions() as db:
        job = await store._job(db, job_id, owner)
        if job.input_json["report"].get("resolutionComplete"):
            raise ConversationError("resolution_state_conflict", "问题状态已经固定", 409)
        job.input_json = {**job.input_json, "report": {**job.input_json["report"],
            "issueResolutions": results, "resolutionComplete": True}}
        await db.commit()
